### PR TITLE
378 attach unit test functionality to output checker plugins

### LINKED ISSUE #339: Add unit tests for output comparisons
# Requested Update

### Description
* Test both good and bad comparisons for the output comparison modules.
* Right now only the good comparisons are tested in the integration tests.

### Background and Motivation
*  [ OPTIONAL: Provide additional detailed information and background related to this issue ]

### Alternative Solutions
*  [ OPTIONAL: If appropriate, list alternative fixes that could be considered]
  
### Environment
*  [ IF APPROPRIATE: Corresponding repository versions ]

### Code to demonstrate issue
*  [ IF APPROPRIATE: Can include steps here to identify/demonstrate bug fix or desired feature ]

### Checklist for Completion
* [ ]

### LINKED ISSUE #374: Create Unit Test Suite for Output Checkers Interface
# Requested Update

### Description

With the new addition of the output_checkers interface, there is a need to add unit testing, to make sure everything is working as expected. This comes from https://github.com/NRLMMD-GEOIPS/geoips/pull/367#discussion_r1358900636, which documents the conversation that brought up this issue. 

### Background and Motivation

Output_checkers are tested in the tests/integration_tests/full_test.sh, however these tests are limited, and we would actually like to test the functionality of output_checkers, rather than just their output.

### Alternative Solutions

Leave the code as is.

### LINKED ISSUE #378: Modify Unit Tests to have some functionality attached to the plugin itself
# Requested Update

### Description

With the addition of unit tests from #374 (see #376, #375), we have begun creating a structure for testing interfaces across GeoIPS and other packages. However, after doing so, we see common pieces of functionality that could be abstracted to a higher level. This issue recommends we attach some functionality to the plugin itself for testing.

### Background and Motivation

Copy and pasting is almost always unnecessary and a bad coding convention. After creating Unit Tests for the Output Checkers interface, we see a need to abstract some of the functionality to the plugins themselves. This issue is a Test to see if this will work, and it will be applied to Output Checker Plugins. 

### Alternative Solutions

Leave the code as is, and deal with the common functions in the Unit Tests. We might have to go this route for simplicity, but it's worth a shot to attach it to the plugins themselves.

### Checklist for Completion
* [ ] Attach Unit Test Functionality to Output Checker Plugins
* [ ] Raise a pytest.xfail if the functionality doesn't exist. (not ready to be tested yet)

### PR DESCRIPTION
**Pull request by evrose54**
# Reviewer Checklist

* [x] Required ***existing tests*** pass (ie full_test.sh, others as appropriate)
* [x] Required ***unit tests*** added and pass for new/modified functionality
* [x] NO REQUIRED ***integration tests*** (explain why not required)
* [x] Required ***documentation*** added for new/modified functionality
* [x] Required ***release notes*** added for new/modified functionality
* [x] NO REQUIRED ***updates to other repos*** (explain why not required)

https://github.com/NRLMMD-GEOIPS/.github/blob/main/.github/review-template.md

# Related Issues

fixes #378 
fixes #374
fixes #339
branched off of #376 

# Testing Instructions
To use this unit test run `pytest test_output_checkers.py` from the directory geoips/tests/unit_tests/plugins/module/output_checkers/

# Summary

With the addition of Unit Tests, we've found there is some commonality between function
signatures. While the functions pertaining to the plugin differ, they use the same
inputs. For this reason, we are going to remove these functions from the Unit Test
itself, and attach them to the plugins under the same nomenclature. If the plugin is
missing these functions, raise a pytest.xfail.

    modified: geoips/tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py
    modified: geoips/geoips/plugins/modules/output_checkers/geotiff.py
    modified: geoips/geoips/plugins/modules/output_checkers/image.py
    modified: geoips/geoips/plugins/modules/output_checkers/netcdf.py
    modified: geoips/geoips/plugins/modules/output_checkers/text.py

# Output
This will test the output checker plugins via pytest. You should get the output shown below.

Screenshot 2023-10-19 at 1 45 24 PM

### CODE REVIEW

**Comment by evrose54** [comment]
Note, do not merge this PR until #376 has been merged.

**Review by mindyls — COMMENTED** [review]
Leaving some notes

File: geoips/plugins/modules/output_checkers/image.py
Comment (evrose54) [reply geoips/plugins/modules/output_checkers/image.py:32]: It uses floats for comparisons, that's just for the file name.

File: geoips/plugins/modules/output_checkers/geotiff.py
Comment (evrose54) [reply geoips/plugins/modules/output_checkers/geotiff.py]: I can change that if wanted, but this now uses resources.files nonetheless.

File: geoips/plugins/modules/output_checkers/image.py
Comment (evrose54) [reply geoips/plugins/modules/output_checkers/image.py]: It's now "get_test_files"

File: geoips/plugins/modules/output_checkers/image.py
Comment (mindyls) [reply geoips/plugins/modules/output_checkers/image.py:37]: Maybe just leave a comment in the code saying those are used in the filenames, and they relate to x, y, z thresholds ?

File: tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py
Comment (mindyls) [inline tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py:25]: @evrose54 @AJThorpe @jsolbrig Do we want to use entry points here before we merge?  So we are not explicitly listing all of the plugins in the unit test infrastructure?  That way it will automatically pick up plugins in other repos, as long as they have the right attributes.

File: geoips/plugins/modules/output_checkers/image.py
Comment (evrose54) [reply geoips/plugins/modules/output_checkers/image.py:37]: I can do that now.

File: tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py
Comment (evrose54) [reply tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py:25]: I'll leave this up to everyone else. I was thinking no, since some output checkers might want different unit tests, but that may not be the case.

Comment (mindyls) [reply tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py:25]: I think we can still implement different unit tests, even if we automatically identify all the output checker plugins here.  If the unit test attributes aren't defined, this will xfail (which I think is fine), and a separate unit test can be implemented.

We *definitely* want all the installed output checkers to be automatically discovered, so that they all can utilitze the base infrastructure without having to explicitly add the plugin name here.

File: geoips/plugins/modules/output_checkers/image.py
Comment (mindyls) [inline geoips/plugins/modules/output_checkers/image.py:33]: This resolved
https://github.com/NRLMMD-GEOIPS/geoips/pull/379#discussion_r1366934180

File: geoips/plugins/modules/output_checkers/image.py
Comment (mindyls) [reply geoips/plugins/modules/output_checkers/image.py]: I think this is more of a philosophical question than anything - should we by convention avoid GEOIPS_PACKAGES_DIR (I think yes, looks like you changed this), and os.path.join?

GEOIPS_PACKAGES_DIR removed here:
https://github.com/NRLMMD-GEOIPS/geoips/pull/379/files#r1373543862

File: geoips/plugins/modules/output_checkers/image.py
Comment (mindyls) [reply geoips/plugins/modules/output_checkers/image.py:37]: Looks good!

File: geoips/plugins/modules/output_checkers/geotiff.py
Comment (mindyls) [inline geoips/plugins/modules/output_checkers/geotiff.py:55]: This resolves https://github.com/NRLMMD-GEOIPS/geoips/pull/379#discussion_r1371991996

File: geoips/plugins/modules/output_checkers/geotiff.py
Comment (mindyls) [reply geoips/plugins/modules/output_checkers/geotiff.py]: I see!  Resolved here:
https://github.com/NRLMMD-GEOIPS/geoips/pull/379/files#r1373547740
Looks good.

File: geoips/plugins/modules/output_checkers/geotiff.py
Comment (mindyls) [inline geoips/plugins/modules/output_checkers/geotiff.py:47]: This resolves https://github.com/NRLMMD-GEOIPS/geoips/pull/379#discussion_r1371993691

File: geoips/plugins/modules/output_checkers/image.py
Comment (mindyls) [reply geoips/plugins/modules/output_checkers/image.py]: Looks good, resolved here: https://github.com/NRLMMD-GEOIPS/geoips/pull/379/files#r1373549085

**Review by mindyls — APPROVED** [review]
This looks good to me once the os.path.join comment has been resolved.  I'm fine either way.

File: tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py
Comment (evrose54) [reply tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py]: I just changed this to be discovered via entry points.

File: tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py
Comment (mindyls) [inline tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py:40]: Can't the output checkers interface be used here directly to list the plugins, then open each one to find it's path?  I think we don't want to rely on the explicit plugins/modules/output_checkers path?

Comment (evrose54) [reply tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py]: That's since been fixed and now uses output_checkers.get_plugins() to discover each available plugin.

File: tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py
Comment (mindyls) [inline tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py:29]: That is much more compact!  

Resolves both
https://github.com/NRLMMD-GEOIPS/geoips/pull/379#discussion_r1373555689
and
https://github.com/NRLMMD-GEOIPS/geoips/pull/379#discussion_r1372495781

File: tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py
Comment (mindyls) [reply tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py]: https://github.com/NRLMMD-GEOIPS/geoips/pull/379/files#r1373579423
Fixed there.

File: tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py
Comment (mindyls) [reply tests/unit_tests/plugins/modules/output_checkers/test_output_checkers.py]: https://github.com/NRLMMD-GEOIPS/geoips/pull/379/files#r1373579423
Fixed there

File: geoips/plugins/modules/output_checkers/text.py
Comment (mindyls) [inline geoips/plugins/modules/output_checkers/text.py:41]: @jsolbrig This is where Evan updated to not use $GEOIPS_PACKAGES_DIR, though perhaps it would be better to just use $GEOIPS_OUTDIRS because if geoips is NOT installed in $GEOIPS_PACKAGES_DIR, ../../ may not be where you expect.

File: geoips/plugins/modules/output_checkers/geotiff.py
Comment (jsolbrig) [reply geoips/plugins/modules/output_checkers/geotiff.py:34]: It would be nice for this to not require `gdal` if possible. Is that something you can look into @evrose54?

File: geoips/plugins/modules/output_checkers/image.py
Comment (jsolbrig) [reply geoips/plugins/modules/output_checkers/image.py]: I think that @AJThorpe used something else in the readers PR. I think it was `gen_test_files` rather than `get_test_files`. We should choose one and be consistent. I kind of lean towards `get_` but would be fine either way.

File: geoips/plugins/modules/output_checkers/text.py
Comment (mindyls) [inline geoips/plugins/modules/output_checkers/text.py:48]: Where is "comp_path" coming from?  I get file does not exist when attempting to run this - should it be auto-generated in the "savedir" directly within this method before attempting to copy/modify it?

Comment (evrose54) [reply geoips/plugins/modules/output_checkers/text.py:48]: Ahh I forgot about this. I'll have to change this for fdeck, text, and geotiff. Should be pretty easy though. Only problem that might occur is for fdeck, since I copied that file from the web.

Comment (mindyls) [reply geoips/plugins/modules/output_checkers/text.py]: Working now!
==================================== 1413 passed, 2 warnings in 23.60s ====================================

File: geoips/plugins/modules/output_checkers/text.py
Comment (mindyls) [reply geoips/plugins/modules/output_checkers/text.py]: Decided to pass in output_path.  See latest file diffs.

File: geoips/plugins/modules/output_checkers/image.py
Comment (mindyls) [reply geoips/plugins/modules/output_checkers/image.py]: Updated to get

File: geoips/plugins/modules/output_checkers/geotiff.py
Comment (mindyls) [reply geoips/plugins/modules/output_checkers/geotiff.py]: Decided to pass in pytest tmp_path

File: geoips/plugins/modules/output_checkers/geotiff.py
Comment (mindyls) [reply geoips/plugins/modules/output_checkers/geotiff.py]: replaced with rasterio

File: geoips/plugins/modules/output_checkers/image.py
Comment (mindyls) [reply geoips/plugins/modules/output_checkers/image.py]: replaced with tmp_path

File: tests/unit_tests/schema/test_plugin_schema/good/product_defaults/interpolator_algorithm.yaml
Comment (mindyls) [reply tests/unit_tests/schema/test_plugin_schema/good/product_defaults/interpolator_algorithm.yaml:1]: Removed test_plugin_schema level

**Review by jsolbrig — APPROVED** [review]

**Pull request merged**
